audio/out/ao_coreaudio_exclusive.c

```diff
@@ -75,13 +75,76 @@ struct priv {
     // Output s16 physical format, float32 virtual format, ac3/dts mpv format
     int spdif_hack;
 
+    struct ao_convert_fmt convert;
+
+    int integer_mode;
+
+    int buffersize;
+
+    int power_saving;
+
+    Float32 IOCycleUsage;
+
     bool changed_mixing;
 
     atomic_bool reload_requested;
 
     uint32_t hw_latency_us;
 };
 
+static int device_property(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+    AudioStreamRangedDescription *formats;
+    size_t n_formats;
+
+    OSStatus err = CA_GET_ARY(p->device, kAudioStreamPropertyAvailablePhysicalFormats,
+                     &formats, &n_formats);
+    if (err != noErr)
+    MP_VERBOSE(ao, "Error: could not get number of device formats\n");
+    int maxbitdepth_physical = 0;
+    int integer_mode_avaliable = 0;
+    int max_mBytesPerPacket = 0;
+    int aligned_high = 0;
+    for (int j = 0; j < n_formats; j++) {
+        AudioStreamBasicDescription *stream_asbd = &formats[j].mFormat;
+        if (stream_asbd->mFormatID == kAudioFormatLinearPCM){ // Exclude spdif format
+            if (stream_asbd->mBitsPerChannel > maxbitdepth_physical)
+                maxbitdepth_physical = stream_asbd->mBitsPerChannel;
+            if (stream_asbd->mFormatFlags & kAudioFormatFlagIsNonMixable)
+                integer_mode_avaliable = 1;
+            if (stream_asbd->mBytesPerPacket > max_mBytesPerPacket)
+                max_mBytesPerPacket = stream_asbd->mBytesPerPacket;
+            if (stream_asbd->mFormatFlags & kLinearPCMFormatFlagIsAlignedHigh)
+                aligned_high = 1;
+            }
+        }
+
+    int device_type = 0;
+    // 20 Bit device support is not tested.
+    if ((integer_mode_avaliable == 1) && (maxbitdepth_physical == 24)
+        && (max_mBytesPerPacket == 8) && (aligned_high == 0)){
+        device_type = 1; // unpacked 24 bit aligned low devce.
+    }else if ((integer_mode_avaliable == 1) && (maxbitdepth_physical == 24)
+        && (max_mBytesPerPacket == 8) && (aligned_high == 1)){
+        device_type = 2; // unpacked 24 bit aligned high devce.
+    }else if ((integer_mode_avaliable == 1) && (maxbitdepth_physical == 24) && (max_mBytesPerPacket == 6)){
+        device_type = 3; // packed 24 Bit device.
+    }else if ((integer_mode_avaliable == 0) && (maxbitdepth_physical == 32)){
+        device_type = 4; // 32 Bit device w/o Integer Mode.
+    }else if ((integer_mode_avaliable == 0) && (maxbitdepth_physical == 24)){
+        device_type = 10; // 24 Bit device w/o Integer Mode.
+    }else if ((integer_mode_avaliable == 0) && (maxbitdepth_physical == 20)){
+        device_type = 11; // 20 Bit device w/o Integer Mode.
+    }else if ((integer_mode_avaliable == 1) && (maxbitdepth_physical == 20)
+        && (max_mBytesPerPacket == 8) && (aligned_high == 0)){
+        device_type = 12; // 20 bit aligned low devce (shown in optical output).
+    }
```

Comment:
I don't think there should be such a device_type integer. They're magical values (and a _lot_ of them), what these magical values mean is widely spread over the entire source, and many of them seem to be used only in 1 place.

It looks like just keeping around integer_mode_avaliable and maxbitdepth_physical as fields would already be a large improvement.

Reply:
Looks like integer_mode_avaliable has a typo, btw.

Reply:
device_type is needed in render_cb_compressed, find_best_format and init, but I agree, we don't have to consider so many different device types at this stage. After all, there are only few 20 Bit or aligned high devices. We could add support later, if anyone raises this issue.

Reply:
What I mean is that you shouldn't enumerate all these cases explicitly. Instead, use the capabilities of the device wherever it matters.

Reply:
I tried, but each time, it needs to inquire`kAudioStreamPropertyAvailablePhysicalFormats` and do the for loop, which add 1-2 sec delay on my computer.

Reply:
Oh come on. Cache it. This is not hard.

Reply:
Sorry, I don't understand what you mean. As I said, I am not a programmer, and this is already the best I can do.

Reply:
This is exactly what nobody ever wants to hear in a code review.